BUG: Handle overlapping line and scatter on the same plot

MartinBraquet opened the pull request:
- [x] closes #61005
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Running the code from the issue above now shows a correct plot.
```python
import datetime
import matplotlib.pyplot as plt
import pandas as pd

datetime_list = [datetime.datetime(year=2025, month=1, day=1, hour=n) for n in range(23)]
y = [n for n in range(23)]
df = pd.DataFrame(columns=['datetime', 'y'])
for i, n in enumerate(datetime_list):
    df.loc[len(df)] = [n, y[i]]
fig, ax = plt.subplots(2, sharex=True)
df.plot.scatter(x='datetime', y='y', ax=ax[0])
df.plot(x='datetime', y='y', ax=ax[1])
```
![image](https://github.com/user-attachments/assets/bd2501ab-4b9c-4507-bb72-2686c27f2455)

MartinBraquet committed:
Refactor time series plotting logic for improved clarity

Extract and streamline time series preparation steps into `prepare_ts_data`, replacing redundant logic across methods. Simplifies axis frequency handling and improves code readability while maintaining functionality.

MartinBraquet committed:
Merge branch 'main' into scatter

MartinBraquet committed:
Add test to validate xtick alignment for scatter and line plots

This test ensures that the x-axis ticks are consistent between scatter and line plots when sharing the same axis. It addresses a potential issue related to GH#61005, verifying proper rendering of datetime x-axis labels.

MartinBraquet committed:
Merge branch 'main' into scatter

MartinBraquet committed:
Fix bug in Series.plot misalignment for line and scatter plots

This resolves an issue where line and scatter plots were not aligned when using Series.plot. The fix ensures proper alignment and improves plot consistency. Refer to issue #61005 for further details.

MartinBraquet committed:
Update scatter plot test to support datetime.time data

Datetime.time is now supported in scatter plots due to added converter implementation in ScatterPlot. Removed the test expecting a TypeError and updated it to validate the new functionality.

MartinBraquet committed:
Refactor handling of x_data in matplotlib plotting.

Simplify and streamline the code by directly assigning x_data from the data variable and replacing the intermediate Series object with a clearer `s` variable. This improves readability and maintains the existing functionality.

MartinBraquet committed:
Move test_scatter_line_xticks from Series to DataFrame tests

Relocated the `test_scatter_line_xticks` test from `test_series.py` to `test_frame.py` for better alignment with DataFrame-specific functionality. This refactor ensures the test resides in the appropriate context based on its usage and focus.

mroeschke commented on pandas/tests/plotting/frame/test_frame.py:
```diff
@@ -840,14 +840,26 @@ def test_plot_scatter_shape(self):
+
+    def test_scatter_line_xticks(self):
+        # GH#61005
+        datetime_list = [datetime(year=2025, month=1, day=1, hour=n) for n in range(3)]
+        df = DataFrame(columns=["datetime", "y"])
+        for i, n in enumerate(datetime_list):
+            df.loc[len(df)] = [n, i]
```
Could you create this in 1 call? i.e. pass a `dict` to `DataFrame`

MartinBraquet replied:
I created it by passing a list

mroeschke commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -54,6 +54,7 @@
 )
 from pandas.core.dtypes.missing import isna
 
+from pandas import Series
```
Can we conditionally import this where used?

MartinBraquet replied:
Done

MartinBraquet committed:
Refactor `prepare_ts_data` to improve type annotations.

Added precise type annotations to the function signature for better clarity and type checking. Replaced `data` with `series` and `kwds` with `kwargs` to enhance readability and consistency.

MartinBraquet committed:
Refactor test_scatter_line_xticks to simplify DataFrame creation

The DataFrame creation in the test has been streamlined for clarity and conciseness by replacing the loop with a list comprehension. This improves code readability and maintains the same functionality.

MartinBraquet committed:
Refactor Series import to optimize scope and maintain consistency

Moved the `Series` import inside relevant function scopes to minimize unnecessary top-level imports and align with existing import patterns. This helps improve code readability and ensures imports are only loaded where needed.

MartinBraquet committed:
`Reorder import statement in _make_plot method`

Moved the import of `Series` within the `_make_plot` method to comply with styling or runtime considerations. This ensures consistency and avoids potential import-related issues.

mroeschke commented:
Thanks @MartinBraquet
